File: Implementation/S2_2504.py
#S2 2504 실패...

string = input()

stack = []

if string[0] == ')':
    print("0")
    exit()
if string[0] == ']':
    print("0")
    exit()
#처음 에러 삭제

for i in range(len(string)):
    tmpsum = 0
    if string[i] == '(':
        stack.append(string[i])
    elif string[i] == '[':
        stack.append(string[i])
    elif string[i] == ')':
        while True:
            tmp = stack.pop()
            if tmp == "[": #에러
                print("0")
                exit()
            elif tmp == "(": #정상
                if tmpsum == 0:
                    stack.append(2)
                else:
                    stack.append(2*tmpsum)
                break
            else:            #정상
                tmpsum += tmp
    elif string[i] == ']':
        while True:
            tmp = stack.pop()
            if tmp == "(": #에러
                print("0")
                exit()
            elif tmp == "[": #정상
                if tmpsum == 0:
                    stack.append(3)
                else:
                    stack.append(3*tmpsum)
                break
            else:            #정상
                tmpsum += tmp
ans = 0
while len(stack) != 0:
    tmp = stack.pop()
    if tmp == '(':
        print("0")
        exit()
    elif tmp == '[':
        print("0")
        exit()
    else:
        ans += tmp
print(ans)

File: DataStructure/Silver/S2_2504.cpp
//https://www.acmicpc.net/problem/10828
//https://www.acmicpc.net/problem/10828
#include <bits/stdc++.h>
#include <string>
using namespace std;
#define endl '\n'

stack<char> s;
int main(void)
{
    cin.tie(0);
    cout.tie(0);
    ios_base::sync_with_stdio(false);

    string bracket;
    cin >> bracket;
    for (int i = 0 ; i < bracket.size() ; ++i)
    {
        cout << bracket[i];
        if (bracket[i] == '(' || bracket[i] == '['){
            s.push(bracket[i]);
            cout << s.top();
        }
        else
        {
            int sum = 0;
            if (!s.size()){
                cout << 0 << endl;
                exit(0);
            }

            while (s.size())
            {
                char tmp = s.top();
                if ((bracket[i] == ')' && tmp == '[') || (bracket[i] == ']' && tmp == '(')){
                    cout << 0 << endl;
                    exit(0);
                }
                if (tmp == '(' || tmp == '[')
                {
                    if (tmp == '(')
                        sum *= 2;
                    else
                        sum *= 3;
                    break;
                }
                else
                    sum += stringstream(tmp);
                s.pop();
            }
            s.push(to_string(sum));
        }
    }
    int sum = 0;
    while (s.size())
    {
        string tmp = s.top();
        if (tmp == "(" || tmp == "["){
            cout << 0 << endl;
            exit(0);
        }
        sum += stringstream(tmp);
    }   
    cout << sum;
}

/*
#include <bits/stdc++.h>
#include <string>
using namespace std;
#define endl '\n'

stack<string> s;
int main(void)
{
    cin.tie(0);
    cout.tie(0);
    ios_base::sync_with_stdio(false);

    string bracket;
    cin >> bracket;
    for (int i = 0 ; i < bracket.size() ; ++i)
    {
        cout << bracket[i];
        if (bracket[i] == '(' || bracket[i] == '['){
            s.push(""+bracket[i]);
            cout << s.top();
        }
        else
        {
            int sum = 0;
            if (!s.size()){
                cout << 0 << endl;
                exit(0);
            }

            while (s.size())
            {
                string tmp = s.top();
                if ((bracket[i] == ')' && tmp == "[") || (bracket[i] == ']' && tmp == "(")){
                    cout << 0 << endl;
                    exit(0);
                }
                if (tmp == "(" || tmp == "[")
                {
                    if (tmp == "(")
                        sum *= 2;
                    else
                        sum *= 3;
                    break;
                }
                else
                    sum += stringstream(tmp);
                s.pop();
            }
            s.push(to_string(sum));
        }
    }
    int sum = 0;
    while (s.size())
    {
        string tmp = s.top();
        if (tmp == "(" || tmp == "["){
            cout << 0 << endl;
            exit(0);
        }
        sum += stringstream(tmp);
    }   
    cout << sum;
}
*/